File: EncryptionDecryption/Substitution.java
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.List;

public class Substitution {
  

  public Substitution(){
    //
  }

  public List<Integer> generateSBox(String seed, int size){
    List<Integer> s_box = new ArrayList<>();
    MessageDigest digest = null;
    try{
      digest = MessageDigest.getInstance("SHA-256");
    }catch(NoSuchAlgorithmException e){
      System.out.println(e.toString());
    }

    for (int i = 0; i < size; i++) {
      byte[] bytes = (seed + (char) i).getBytes();
      byte[] hash = digest.digest(bytes);
      String hashString = bytesToHex(hash);
      int candidateElmt = Integer.parseInt(hashString.substring(0, 2), 16);
      if (!s_box.contains(candidateElmt)) {
          s_box.add(candidateElmt);
      }
    }
    
    for (int i = 0; i < size; i++) {
        if (!s_box.contains(i)) {
            s_box.add(i);
        }
    }

    return s_box;
  }

  private String bytesToHex(byte[] bytes) {
    StringBuilder sb = new StringBuilder();
    for (byte b : bytes) {
        sb.append(String.format("%02x", b));
    }
    return sb.toString();
  }

  public List<Integer> inverseSBox(List<Integer> s_box){
    List<Integer> inverse = new ArrayList<>();

    for(int i = 0; i < s_box.size(); i++){
      inverse.add(s_box.indexOf(i));
    }

    return inverse;
  }

  public byte[] substitute(byte[] string_byte, List<Integer> s_box){
    byte[] result = new byte[string_byte.length];
    int i = 0;
    for(byte s :  string_byte){
      result[i] = s_box.get(s & 0xff).byteValue();
      i++;
    }

    return result;
  }

  public byte[] reverse(byte[] string_byte, List<Integer> s_box){
    byte[] result = new byte[string_byte.length];
    int i = 0;
    for(byte s :  string_byte){
      result[i] = (byte) s_box.indexOf(s & 0xff);
      i++;
    }

    return result;
  }
}

File: EncryptionDecryption/Substitution.py
import hashlib

def generateSBox(seed : str, size : int) -> list:
    """
        Generate S-Box menggunakan fungsi hash SHA-256.
        Generate dilakukan juga dengan memastikan bahwa
        element S-Box bernilai unik
    """
    s_box = []
    for i in range(size):
        hash = hashlib.sha256(seed.encode() + bytes([i])).hexdigest()
        candidate_elmt = format(int(hash[:2], 16), '02x')
        if (candidate_elmt not in s_box):
            s_box.append(candidate_elmt)

    for i in range(size):
        candidate_elmt = format(i,'02x')
        if(candidate_elmt not in s_box):
            s_box.append(candidate_elmt)
    # print(len(s_box) == len(set(s_box)))
    return s_box

def inverseSBox(s_box : list):
    '''
        S-Box inverse dibuat dengan membuat list 
        yang berisi indeks pada S-Box asli
    '''
    s_box_inverse = []
    for i in range(len(s_box)):
        s_box_inverse.append(s_box.index(format(i,'02x')))
    
    return s_box_inverse

def substitute(string_byte, s_box:list):
    result = b''
    for b in string_byte:
        result += bytes.fromhex(s_box[b])
    return result

def reverse(string_byte, s_box:list):
    result = b''
    for b in string_byte:
        result += bytes([s_box.index(format(b,'02x'))])
        #result += bytes.fromhex(format(s_box.index(bytes(b).hex()),'02x'))
    return result
    

# s_box = generateSBox("H-2 Menuju UTS Semangat", 256)
# s_box_inverse = inverseSBox(s_box)
# test_text = b"Berserah diri keawedawedawedpada Tuhan"

# x = substitute(test_text,s_box)
# y = reverse(x, s_box)

'''
    Substitusi dilakukan dengan iterasi nilai byte pada teks.
    Dan digunakan sebagai indeks untuk mendapatkan nilai pada
    S-Box
'''
# substituted_block = [s_box[b] for b in test_text]
# print(s_box)
# print(len(s_box))
# print(s_box.index(215))
# print(s_box_inverse)